ark/jsonschema/json.ts

```diff
@@ -0,0 +1,91 @@
+import type { JsonSchemaOrBoolean } from "@ark/schema"
+import { printable, throwParseError } from "@ark/util"
+import { type, type JsonSchema } from "arktype"
+import { parseJsonSchemaAnyKeywords } from "./any.ts"
+import { validateJsonSchemaArray } from "./array.ts"
+import { parseJsonSchemaCompositionKeywords } from "./composition.ts"
+import { validateJsonSchemaNumber } from "./number.ts"
+import { validateJsonSchemaObject } from "./object.ts"
+import { JsonSchemaScope } from "./scope.ts"
+import { validateJsonSchemaString } from "./string.ts"
+
+export const innerParseJsonSchema = JsonSchemaScope.Schema.pipe(
+	(jsonSchema: JsonSchemaOrBoolean): type.Any => {
+		if (typeof jsonSchema === "boolean") {
+			if (jsonSchema) return JsonSchemaScope.Json
+			else return type("never") // No runtime value ever passes validation for JSON schema of 'false'
+		}
+
+		if (Array.isArray(jsonSchema)) {
+			return (
+				parseJsonSchemaCompositionKeywords({ anyOf: jsonSchema }) ??
+				throwParseError(
+					"Failed to convert root array of JSON Schemas to an anyOf schema"
+				)
+			)
+		}
+
+		const constAndOrEnumValidator = parseJsonSchemaAnyKeywords(
+			jsonSchema as JsonSchema
+		)
+		const compositionValidator = parseJsonSchemaCompositionKeywords(
+			jsonSchema as JsonSchema
+		)
+
+		const preTypeValidator: type.Any | undefined =
+			constAndOrEnumValidator ?
+				compositionValidator ? compositionValidator.and(constAndOrEnumValidator)
+				:	constAndOrEnumValidator
+			:	compositionValidator
+
+		if ("type" in jsonSchema) {
+			let typeValidator: type.Any
+
+			if (Array.isArray(jsonSchema.type)) {
+				typeValidator =
+					parseJsonSchemaCompositionKeywords({
+						anyOf: jsonSchema.type.map(t => ({ type: t }))
+					}) ??
+					throwParseError(
+						"Failed to convert array of JSON Schemas types to an anyOf schema"
+					)
+			} else {
+				const jsonSchemaType = jsonSchema.type as JsonSchema.TypeName
+				switch (jsonSchemaType) {
```

Comment:
Man would be nice if I had released pattern matching for this- would have been a great opportunity to dogfood!

You could still try and convert some over if you want, but not at all required.

Reply:
I'll wait for the `match` functionality to be in main, and then see if I can convert some of this over -- it would certainly be nice to take advantage of it!

Reply:
Done in bdc394e65!